### PR TITLE
qr testing mode, fixed qr pure address and EIP681 issues

### LINKED ISSUE #7214: Add custom scheme support for ethereum-*
### User Story

As a user, I want to have ethereum links (EIP [681](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-681.md)) redirect to status wallet.
Also clicking on such link in external apps should open status wallet.

*Type*: Feature

### LINKED ISSUE #8150: EIP681 `value`, `chain_id`, `gasLimit` and `gasPrice` are not supported
### Description

*Type*: Bug
*Summary*: Some of the parameters mentioned in [EIP681 spec](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-681.md) are not supported in Status app.

1)  Links without `function_name` are not resolved, i.e. `ethereum:0xfb6916095ca1df60bb79Ce92ce3ea74c37c5d359?value=2.014e18` so when scanning URL with this link, got an error 
Monosnap 2019-05-01 11-19-13
So I didn't find a way to check `value` for ETH

2) If using ERC20 tokens as parameters, value is not resolved, i.e. after scanning `ethereum:0x744d70fdbe2ba4cf95131626614a1763df805b9e/transfer?address=0x8e23ee67d1332ad560396262c48ffbb01f93d052&uint256=1` should sent 1 SNT as far as I understand.
In fact after scanning QR with link I've got:
photo_2019-05-01 11 27 14

So the asset was changed to tokens, but the value is not displayed.

3) Parameters `gasLimit` and `gasPrice` are not resolved (i.e. in `ethereum:0x744d70fdbe2ba4cf95131626614a1763df805b9e/transfer?address=0x8e23ee67d1332ad560396262c48ffbb01f93d052&value=1&gasLimit=100&gasPrice=1`
However `gas` is resolved, as it mentioned in [spec](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-681.md)

4) `chain_id` is not resolved
atattempt to scan `ethereum:0x9810fc231ca9dabe30da595763a7b81a215aaceb/transfer?chain_id=3&address=0x8e23ee67d1332ad560396262c48ffbb01f93d052&value=1` on Ropsten got an error:

#### Expected behavior
all paramters are pasted on "Send transaction" screen

#### Actual behavior
mentioned in description

### Reproduction

- Make valid links according to  [EIP681 spec](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-681.md)
- Make a QR with any QR code genarator
- Open Status
- Tap on + > Scan QR

- Status version: nightly 08/05/2019
[comment]: # (Android or iOS.)
- Operating System: Android, iOS

### LINKED ISSUE #9183: URL Format for Transaction Requests (EIP681) is poorly supported
EIP681 is supported only in QRCodes and the specification is partially supported.

### EIP681 - URL Format for Transaction Requests
https://eips.ethereum.org/EIPS/eip-681
A link that pop up a transaction in the wallet.
**In Status.im:** Useful as chat message and for dapp support.

#### Examples 
Donate 0.314 SNT (0x744d70fdbe2ba4cf95131626614a1763df805b9e) to 3esmit's address (0x3d597789ea16054a084ac84ce87f50df9198f415), no ens:
`ethereum:0x744d70fdbe2ba4cf95131626614a1763df805b9e/transfer?address=0x3d597789ea16054a084ac84ce87f50df9198f415&uint256=314e17`

Donate 0.1337 SNT to 3esmit.stateofus.eth using ens for token (snt.thetoken.eth) and for destination address:
`ethereum:snt.thetoken.eth/transfer?address=3esmit.stateofus.eth&uint256=1337e17`

Donate 0.011 SNT to 3esmit.stateofus.eth using ens only for destination address:
`ethereum:0x744d70fdbe2ba4cf95131626614a1763df805b9e/transfer?address=3esmit.stateofus.eth&uint256=11e16`

Donate 2.014 ETH to Ethereum Foundation with manual gas settings:
`ethereum:0xfb6916095ca1df60bb79Ce92ce3ea74c37c5d359?value=2.014e18&gasPrice=1000000000&gasLimit=100000`

#### Related 
#7214 
#7325 
#8150 

#### Code
https://github.com/status-im/status-react/blob/develop/src/status_im/ethereum/eip681.cljs

### LINKED ISSUE #9580: Amount and asset are reset when scanning ETH address
# Bug Report

## Problem

Usually users set amount, then asset, then select From and To when sending transaction.
If in field TO scan ethereum address (i.e. `ethereum: 0xcf2272205cc0cf96Cfbb9Dd740BD681D1E86901E` - all previous parameters except From will be reset. 
It is very common flow so issue is very visible.

#### Expected behavior
 Asset and Amount are preserved, can sign transaction

#### Actual behavior
 Asset and Amount are reset, need to change them second time
screencast 2019-12-04 15-54-15 mp4 2019-12-04 15-55-28

### Reproduction

[comment]: # (Describe how we can replicate the bug step by step.)

- Open Status
- Create multiaccount > Send
- Change Asset and Amount
- Scan valid ethereum address

### Additional Information
- Status version: nightly 04.12.2019
- Operating System: Android- iOS

### LINKED ISSUE #9872: "Send transaction" screen opens after scanning invalid QR in wallet
# Bug Report

## Problem

When scanning invalid address, after dismissing popup with error, "Send transaction" screen is opened without default values

#### Expected behavior
no new screen

#### Actual behavior
photo_2020-01-21 18 01 15

### Reproduction

- Open Status
- Go to wallet > scan QR in the top 
- Scan invalid QR

### Additional Information

- Status version: nightly 21/01/2020
- Operating System: Android, iOS

### LINKED ISSUE #10093: Value is not supported in EIP681 links
# Bug Report

## Problem

At attempt to scan QRs like `ethereum:pay-0xd21DB0e43048AcB94f428eD61dC244c82f1ff2a8/transfer?address=0xd21DB0e43048AcB94f428eD61dC244c82f1ff2a8&uint256=1e2` or `ethereum:0x3d597789ea16054a084ac84ce87f50df9198f415?value=1` value is not set.

Mentioned in #9380 and in #9563 

#### Expected behavior
can see value in 'Amount' field after scanning

#### Actual behavior
Send transaction bottom sheet is opened; recepient and assets are set; value is not set

### Reproduction

[comment]: # (Describe how we can replicate the bug step by step.)

- Open Status
- Go to wallet
- Scan QR
`ethereum:0x3d597789ea16054a084ac84ce87f50df9198f415?value=1e15`
QR Code Generator | Create Your Free QR Codes 2020-02-27 16-12-16

or 
`ethereum:pay-0xe41d2489571d322189246dafa5ebde1f4699f498/transfer?address=0xd21DB0e43048AcB94f428eD61dC244c82f1ff2a8&uint256=2e1`
EIP-681 - Ethereum transaction link generator 2020-02-27 16-17-06

### Additional Information

- Status version: nightly 27/02/2020
- Operating System: Android, iOS

### LINKED ISSUE #10107: Wallet QRCode Scanner says invalid address for valid address scanned
# Bug Report

## Problem

Not all sites use EIP-681 for expressing payment requests, or the QRCode is used only to share ethereum address. For example in Etherscan and many crypto payment systems.

image

How to reproduce:

1. Go to Wallet
2. Open the desired wallet
3. Tap Send
4. Enter some value
5. Scan the QRCode of address, as example above. 

image

#### Expected behavior

When only an address is entered, it should be filled in the address, otherwise it must be following ERC-681

#### Actual behavior

Status don't know what to do with scanned address.

### PR DESCRIPTION
**Pull request by flexsurfer**
fixes #10107
fixes #10093
fixes #9183
fixes #8150
fixes #9580 
fixes #9872

removed chain usage for tokens 

For local and e2e builds, qr scan testing mode implemented, so now we can add e2e for qr codes

Tested with this generator https://brunobar79.github.io/eip681-link-generator/#

FOR QA: all tokens related functionality should be tested, also QR codes , especially value

### CODE REVIEW

File: src/status_im/ethereum/eip681.cljs [hunk 1]
```diff
@@ -51,26 +51,28 @@
...
-              (when-let [arguments (parse-arguments function-name query)]
-                (let [contract-address (get-in arguments [:function-arguments :address])]
-                  (if-not (or (not contract-address) (or (ens/is-valid-eth-name? contract-address) (ethereum/address? contract-address)))
-                    nil
-                    (merge {:address address
-                            :chain-id (if chain-id
-                                        (js/parseInt chain-id)
-                                        (ethereum/chain-keyword->chain-id :mainnet))}
-                           arguments)))))))))))
+    (if (ethereum/address? s)
```
Comment (flexsurfer) [inline src/status_im/ethereum/eip681.cljs:54]: only this one small change in this function, added address condition for pure addreses

**Review by cammellos — APPROVED** [review]

**Comment by jakubgs** [comment]
I just tested the QR code and it's using just the chat key, not the full URL including `join.status.im`.

Are we sure we want to do it that way? With that the QR code is useless for anyone that doesn't have Status installed, and if it was a full URL it would be usable for anyone.

**Comment by jakubgs** [comment]
Could you also take a look at #10306 as part of this? Might affect address scanner.

**Comment by flexsurfer** [comment]
this PR is only for status wallet qr scanner. Chat scanner and mailserver scanner are separate so should be done in separate PRs

**Comment by jakubgs** [comment]
That makes sense, but does the wallet scanner handle whitespace/newlines fine?

**Comment by flexsurfer** [comment]
Good question :) i need to take a look

**Comment by flexsurfer** [comment]
@Serhy @churik any chance ? :)

**Comment by churik** [comment]
@flexsurfer 
### ISSUE 1: In case of error (i.e. wrong network id in the link) empty bottom sheet is opened
Link: ethereum:0x4470BB87d77b963A013DB939BE332f927f2b992e@5/transfer?address=0xd21DB0e43048AcB94f428eD61dC244c82f1ff2a8&uint256=1e-5

photo_2020-04-10 15 02 02
photo_2020-04-10 15 02 00
Tested on Android.

### ISSUE 2:  Sometimes after scanning QR-codes from Recipient -> Scan QR app is freezing with opened bottom sheet 
Link: ethereum:0x4470BB87d77b963A013DB939BE332f927f2b992e@5/transfer?address=0xd21DB0e43048AcB94f428eD61dC244c82f1ff2a8&uint256=1e-5

Steps to reproduce are scan link from Wallet -> QR code and then from Wallet -> Send Transaction -> Choose recipient -> Scan QR code and repeat until get unresponsive app (in my case 3-4 times)
At ~10 attempts got 2 freezing.

Full logcat with 2 freezing: 
[logcat.log](https://github.com/status-im/status-react/files/4461661/logcat.log)
Reproducible on IOS and Android

UPD: it does not depend on the place of scanning (Wallet -> QR or Send Transaction -> Recipient -> Scan QR) or link itself - can be reproduced with any link when scanning QR several times. Sometimes it is reproducible from the 1 time.


Testing is in progress.

**Comment by churik** [comment]
### ISSUE 3: `gasPrice` and `gasLimit` are still not supported

Link: 
`ethereum:0xcf2272205cc0cf96Cfbb9Dd740BD681D1E86901E?value=1e17&gasPrice=1000000000&gasLimit=100000`

At attempt to scan it and navigate to "Next" custom gas price / gas limit values are set.

### ISSUE 4: Links with Ethereum Foundation address are not scannable
Link (from official https://github.com/ethereum/EIPs/blob/master/EIPS/eip-681.md): 
ethereum:0xfb6916095ca1df60bb79Ce92ce3ea74c37c5d359?value=2.014e18 
Invalid address error at attempt to scan.

### ISSUE 5: When scanning QR code with ENS name all parameters except address are ignored
Link (20 SNT to ENS):
` ethereum:0x744d70fdbe2ba4cf95131626614a1763df805b9e/transfer?address=tanyatest1.eth&uint256=2e1`
On scanning:
photo_2020-04-10 17 38 51

### ISSUE 5: ENS name for tokens is not resolved
At attempt to scan link
`ethereum:snt.thetoken.eth/transfer?address=0xcf2272205cc0cf96Cfbb9Dd740BD681D1E86901E&uint256=2e1` with ENS name for tokens got Invalid address error

**Comment by churik** [comment]
Thank you for making possibility for e2e to scan QR!
Works awesome 👍

**Comment by flexsurfer** [comment]
- [x] ISSUE 1: In case of error (i.e. wrong network id in the link) empty bottom sheet is opened
- [x] ISSUE 2: Sometimes after scanning QR-codes from Recipient -> Scan QR app is freezing with opened bottom sheet **(isn't reproducible, probably not related to this PR, could be done in separate issue)**
- [x] ISSUE 3: gasPrice and gasLimit are still not supported
- [x] ISSUE 4: Links with Ethereum Foundation address are not scannable **(address isn't valid in the example)**
- [x] ISSUE 5: When scanning QR code with ENS name all parameters except address are ignored **(isn't reproducible)**
- [ ] ISSUE 5: ENS name for tokens is not resolved **(not in the scope of this PR, in separate issue please)**

**Comment by churik** [comment]
also fixes #9872

**Comment by churik** [comment]
Tested on IOS13 and Android 10:
- payment URLs
- URLs with differents amount of tokens and ETH
- URLs with custom gas settings
- URLs with ENS for receiver
- URLs with chain_id
- error handling of the wrong chain_id, wrong URL, wrong ENS, too precise amount, too big amount, too big gas
- scanning EIP681 URLs with camera and opening in Status when app is on background and app is closed
- scanning valid and invalid QR in + > Start new chat
- scanning QR from To > Enter recipient address - only address is resolved, as expected.

All mentioned issues are resolved, only ENS name for tokens is not resolved (will be added separately).

AWESOME 👍

**Pull request merged**